intern/openmesheffect/blender/intern/mfxCallbacks.cpp

```diff
@@ -131,33 +208,164 @@ OfxStatus before_mesh_get(OfxHost *host, OfxMeshHandle ofx_mesh) {
   OfxPropertySetHandle vertpoint_attrib;
   MFX_CHECK(mes->meshGetAttribute(
       ofx_mesh, kOfxMeshAttribVertex, kOfxMeshAttribVertexPoint, &vertpoint_attrib));
-  
-  MFX_CHECK(ps->propSetInt(vertpoint_attrib, kOfxMeshAttribPropIsOwner, 0, 0));
-  MFX_CHECK(ps->propSetPointer(
-      vertpoint_attrib, kOfxMeshAttribPropData, 0, (void *)&blender_mesh->mloop[0].v));
-  MFX_CHECK(ps->propSetInt(vertpoint_attrib, kOfxMeshAttribPropStride, 0, sizeof(MLoop)));
+
+  if (0 == blender_loose_edge_count) {
+    // use host buffers, kOfxMeshPropNoLooseEdge optimization
+    MFX_CHECK(ps->propSetInt(vertpoint_attrib, kOfxMeshAttribPropIsOwner, 0, 0));
+    MFX_CHECK(ps->propSetPointer(
+        vertpoint_attrib, kOfxMeshAttribPropData, 0, (void *)&blender_mesh->mloop[0].v));
+    MFX_CHECK(ps->propSetInt(vertpoint_attrib, kOfxMeshAttribPropStride, 0, sizeof(MLoop)));
+  } else {
+    // request new buffer, we need to append new vertices for loose edges
+    MFX_CHECK(ps->propSetInt(vertpoint_attrib, kOfxMeshAttribPropIsOwner, 0, 1));
+  }
 
   // Face count
   OfxPropertySetHandle facecounts_attrib;
   MFX_CHECK(mes->meshGetAttribute(
       ofx_mesh, kOfxMeshAttribFace, kOfxMeshAttribFaceCounts, &facecounts_attrib));
 
-  MFX_CHECK(ps->propSetInt(facecounts_attrib, kOfxMeshAttribPropIsOwner, 0, 0));
-  MFX_CHECK(ps->propSetPointer(
-      facecounts_attrib, kOfxMeshAttribPropData, 0, (void *)&blender_mesh->mpoly[0].totloop));
-  MFX_CHECK(ps->propSetInt(facecounts_attrib, kOfxMeshAttribPropStride, 0, sizeof(MPoly)));
+  if (-1 != ofx_constant_face_count) {
+    // no buffer, kOfxMeshPropConstantFaceCount optimization
+    MFX_CHECK(ps->propSetInt(facecounts_attrib, kOfxMeshAttribPropIsOwner, 0, 0));
+    MFX_CHECK(ps->propSetPointer(facecounts_attrib, kOfxMeshAttribPropData, 0, NULL));
+    MFX_CHECK(ps->propSetInt(facecounts_attrib, kOfxMeshAttribPropStride, 0, sizeof(int)));
+  } else if (0 == blender_loose_edge_count) {
+    // use host buffers, kOfxMeshPropNoLooseEdge optimization
+    MFX_CHECK(ps->propSetInt(facecounts_attrib, kOfxMeshAttribPropIsOwner, 0, 0));
+    MFX_CHECK(ps->propSetPointer(
+        facecounts_attrib, kOfxMeshAttribPropData, 0, (void *)&blender_mesh->mpoly[0].totloop));
+    MFX_CHECK(ps->propSetInt(facecounts_attrib, kOfxMeshAttribPropStride, 0, sizeof(MPoly)));
+  } else {
+    // request new buffer, we need to append new faces for loose edges
+    MFX_CHECK(ps->propSetInt(facecounts_attrib, kOfxMeshAttribPropIsOwner, 0, 1));
+  }
 
+  // finished adding attributes, allocate any requested buffers
   MFX_CHECK(mes->meshAlloc(ofx_mesh));
 
+  // loose edge cleanup
+  if (blender_loose_edge_count > 0) {
+    // check that meshAlloc() gave us contiguous buffers
```

Comment:
Is it that important to have contiguous buffers? If so, we should specify it in a comment in the meshAlloc implementation

Reply:
Tight packing probably has some performance benefit, but it's mostly to make the code a bit easier to read/write (avoid the full "casting pattern"). It's stating that `meshAlloc()` will allocate buffers via malloc/new, which will be contiguous; `meshAlloc()` would have to go out of its way to do something different. I'm asserting it just to be sure.

Technically it's an implementation detail of this host's `meshAlloc()`, mentioning it in its docstring is probably a good idea :) It can be cleared up when it's time to implement arbitrary attributes, which will require some major surgery in the callback code anyway.

Reply:
Yes it's only an internal choice but I'll add it to the docstring cause eventually we might want to try and aligning some data buffers for performances, e.g. allocating 3 component float attributes with a 4*32bit stride for whatever paralelization performance issue. I don't have precise plans about this, but it may come up one day and that day we'll have forgotten that somewhere in the code we made the assumption that things were contiguous (fortunately, you've `assert`ed it so it won't trigger too much of weird bugs ^^)

Reply:
Good point about the alignment, that totally slipped my mind :)